ncbi-submit/metadata_client.py

```diff
@@ -0,0 +1,282 @@
+        metadata = bio_sample_folder['metadata']
+
+        for attribute in metadata:
+            if not (attribute.get('attr') and attribute.get('value')):
+                continue
+
+            attr = attribute['attr']
+            value = attribute['value']
+            if attr in self.bio_sample_reserved_attributes:
+                if attr in bio_sample:
+                    raise Exception("Duplicate '{0}' attribute found in Bio Sample metadata.\nValues:\n{1}\n{2}".format(attr, bio_sample[attr], value))
+
+                bio_sample[attr] = value
+            else:
+                bio_sample['attributes'].append({'name': attr, 'value': value})
+                if attr in self.bio_sample_dup_attributes:
+                    # This attribute needs to be duplicated outside the attributes node as well.
+                    bio_sample[attr] = value
+
+        bio_sample['libraries'] = [self._parse_library_metadata(bio_sample['sample_id'], bio_sample['name'], library)
+                                   for library in bio_sample_folder['folders']]
+
+        return bio_sample
+
+    def _parse_library_metadata(self, sample_id, bio_sample_name, library_folder):
+        if not library_folder.get('path'):
+            raise Exception("Could not find Bio Sample Library folder path")
+
+        library_name = os.path.join(bio_sample_name, os.path.basename(library_folder['path']))
+
+        if not library_folder.get('id'):
+            raise Exception("Could not find Bio Sample Library ID: {0}".format(library_name))
+        if not library_folder.get('metadata'):
+            raise Exception("Could not load Bio Sample Library metadata: {0}".format(library_name))
+        if self.require_files and not library_folder.get('files'):
+            raise Exception("Could not find Bio Sample Library file metadata: {0}".format(library_name))
+
+        library = {"library_id": library_folder['id'],
+                   "name": library_name,
+                   "sample_id": sample_id,
+                   "attributes": []}
+        metadata = library_folder['metadata']
+
+        # Initialize categorized attributes.
+        for category_name in set(self.library_categorized_attributes.values()):
+            library[category_name] = {}
+
+        for attribute in metadata:
+            if not (attribute.get('attr') and attribute.get('value')):
+                continue
+
+            attr = attribute['attr']
+            value = attribute['value']
+            if attr in self.library_reserved_attributes:
+                if attr in library:
+                    raise Exception("Duplicate '{0}' attribute found in Bio Sample Library metadata.\nValues:\n{1}\n{2}".format(attr, library[attr], value))
+
+                library[attr] = value
+            elif attr in self.library_categorized_attributes:
+                category_name = self.library_reserved_attributes[attr]
+                library[category_name][attr] = value
```

Comment:
OK, I looked up what `py:attrs` will do in the XML template, so if my other assumptions are correct, will this code add attributes to that node like the following?
```
<GenomeRepresentation genome_representation_description="some-value">library.genome_representation</GenomeRepresentation>
```
But again, should the `category_name` be coming from `self.library_categorized_attributes[attr]` instead of `self.library_reserved_attributes[attr]`?